File: coding_builds/lib/coding_builds.ex
defmodule CodingBuilds do
  @doc """
  # coding builds
  # GET https://cmcmus.coding.net/api/enterprise/cmcmus/workbench/builds?page=1&pageSize=30
  """
  def main(_args \\ []) do
    token = Application.fetch_env!(:coding_builds, :token)

    url = "https://cmcmus.coding.net/api/enterprise/cmcmus/workbench/builds"

    # ====== Headers ======
    headers = [
      {"Authorization", "token #{token}"}
    ]

    # ====== Query Params ======
    params = [
      {"page", "1"},
      {"pageSize", "30"}
    ]

    HTTPoison.start()

    case HTTPoison.get(url, headers, params: params) do
      {:ok, %HTTPoison.Response{status_code: _status_code, body: body}} ->
        pretty_response(body)

      {:error, error = %HTTPoison.Error{reason: reason}} ->
        IO.puts("Request failed: #{reason}")

        error
    end
  end

  defp pretty_response(body) do
    result = Jason.decode!(body)

    case result["code"] do
      0 ->
        builds = result["data"]["list"]

        pretty_string =
          builds
          |> Enum.map(fn build -> "number #{build["number"]}" end)
          |> Enum.join("\n")

        IO.puts(pretty_string)

      _ ->
        IO.puts(body)
    end
  end
end

File: coding_builds.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'net/http'
require 'net/https'
require 'json'
# coding builds (GET )
def send_request(token)
  uri = URI('https://cmcmus.coding.net/api/enterprise/cmcmus/workbench/builds?page=1&pageSize=10')

  # Create client
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  http.verify_mode = OpenSSL::SSL::VERIFY_PEER

  # Create Request
  req = Net::HTTP::Get.new(uri)
  # Add headers
  req.add_field 'Authorization', "token #{token}"

  # Fetch Request
  res = http.request(req)
  res.body
rescue StandardError => e
  puts "HTTP Request failed (#{e.message})"
end

def pretty_status_emoji(status)
  if status == 'SUCCEED'
    '✅'
  else
    '❌'
  end
end

def pretty_line(build)
  job = build['jobName']
  status = build['status']
  status_emoji = pretty_status_emoji(status)

  "#{status_emoji} #{build['cause']} #{job} #{build['number']}"
end

def pretty_first_line(build)
  status = build['status']
  status_emoji = pretty_status_emoji(status)

  "Build: #{status_emoji} #{build['number']}"
end

token = File.read("#{__dir__}/.coding.token")
body = send_request(token)
result = JSON.parse(body)

builds = result['data']['list']

puts pretty_first_line(builds.first)
puts '---'
puts builds.map { |build| pretty_line(build) }.join("\n")
puts '---'
puts 'Refresh... | refresh=true'
puts '---'
puts 'Dashboard Web | href=https://cmcmus.coding.net/user'
